src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,211 @@
+ * return std::nullopt if no lossless conversion was possible.
+ *      lossless_cast<Target>(Source)
+ *      lossless_variant_cast<Target>(AllTypeVariant)
+ *      lossless_variant_cast(Source, DataType)
+ *
+ * "Lossless" means just that: No information must be lost during casting. Typical lossful operations would be
+ *      - float-to-int casts with the float having a non-zero fractional part
+ *      - long-to-int casts where the long was > MAX_INT
+ *      - string-to-int where the string contained no number, or a number > MAX_INT.
+ *      - double-to-float where the significand of the double has bits set that cannot be represented in float
+ *
+ * Whenever possible you should use these casts over "comfortable" casts such as static_cast or boost::lexical_cast.
+ * Those might incur information loss and might thus introduce bugs. See #1306, #1525.
+ */
+
+namespace opossum {
+
+// Identity
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<Target, std::decay_t<Source>>, std::optional<Target>> lossless_cast(Source&& source) {
+  return std::forward<Source>(source);
+}
+
+// Int64 to Int32
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<int64_t, Source> && std::is_same_v<int32_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  if (source < std::numeric_limits<int32_t>::min() || source > std::numeric_limits<int32_t>::max()) {
+    return std::nullopt;
+  } else {
+    return static_cast<Target>(source);
+  }
+}
+
+// Int32 to Int64
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<int32_t, Source> && std::is_same_v<int64_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return static_cast<Target>(source);
+}
+
+// NULL to anything but NULL
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<NullValue, Source> && !std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// Anything but NULL to NULL
+template <typename Target, typename Source>
+std::enable_if_t<!std::is_same_v<NullValue, Source> && std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// String to Integral
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<pmr_string, Source> && std::is_integral_v<Target>, std::optional<Target>> lossless_cast(
+    const Source& source) {
+  static_assert(std::is_same_v<int32_t, Target> || std::is_same_v<int64_t, Target>, "Expected int32_t or int64_t");
+
+  Target result{};
+
+  if (boost::conversion::detail::try_lexical_convert(source, result)) {
+    return result;
+  } else {
+    return std::nullopt;
+  }
+}
+
+// String to Floating Point
+// NOT SUPPORTED: Some strings (e.g., "5.5") have lossless float representations, others (e.g., "5.3") do not. Allowing
```

Comment:
"All over the place"? And now I am left guessing what actually bothers you. 

(I all-capped "NOT SUPPORTED" because I judged the following comment to be more important than the others in this file.)

Reply:
"String" vs "strings"
"Floating Point" vs "floating point" vs "float"

Reply:
hopefully better now. "floating point" (which includes double and float) and "float" is something different intentionally, btw.

Reply:
Absolutely.

> "floating point" (which includes double and float) and "float" is something different intentionally, btw.

Yes, which is why the case was more confusing.